SetPointManager Refactor

amirroth opened the pull request:
Object-oriented refactor of the `SetPointManager` module. The original module had a class hierarchy but instances of each class were contained in separate arrays resulting in a significant amount of redundant code. The new implementation uses a single array of pointers to objects of all `SetPointManager` sub-classes, enabling significant code reuse.  Overall, almost 6,000 lines of code (65%) were removed from `SetPointManager.[cc|hh]`. This was an especially redundant module and so savings of this magnitude may not be possible elsewhere, but savings of about half that should be. We'll have to see.

Some highlights:
- Changed input processing approach from IDF-based to epJSON-based, this enabled significant code savings in the `GetInput` routine. 
- In the new `GetInput` routine, there is code that is guarded by a `PRESERVE_IDF_ORDER` macro. This establishes a template for preserving IDF order in epJSON-based input code. The approach is the same as the one used in `getObjectItem` (thanks @mjwitte).
- Streamlined default value handling `getRealFieldValue` and `getAlphaFieldValue`. 
- Moved some state variables to local.
- Added duplicate name check and fixed one test file to comply.

CI is green except for a few diffs in some `.bnd` files. These are caused by the fact that SetPointManager input data is now read and processed only once. In some simulations it used to be processed twice, once in `InternalGains` and then again later in `ManageHVAC`. Double processing caused differences in the number of times `LoopNode` objects were referenced. I suspect that this is also why there wasn't a duplicate name check. When input is processed twice, every name is a duplicate.

amirroth committed:
Initial refactor

amirroth committed:
Merge remote-tracking branch 'origin/develop' into SPM

amirroth committed:
Fix some CI errors

amirroth committed:
Merge remote-tracking branch 'origin/develop' into SPM

amirroth committed:
Fix a few more CI issues

amirroth committed:
Yet more CI fixes

amirroth committed:
Whittling away

amirroth committed:
Whittle whittle

amirroth committed:
Preserve node creation order (and my sanity)

amirroth committed:
Fix MZMinHum and IdealCondenserEnteringTemp managers

amirroth committed:
Fix SZCooling manager

amirroth committed:
Merge remote-tracking branch 'origin/develop' into SPM

amirroth committed:
Fix SZReheat and FollowSystemNode inputs

amirroth committed:
Fix FollowSysNode and SZMaxHum managers

amirroth committed:
Fix ReturnWaterTemp manager

amirroth committed:
Fix MZHeating and MZCooling managers

amirroth committed:
Preserve IDF object order within object type

amirroth committed:
Merge remote-tracking branch 'origin/develop' into SPM

amirroth committed:
Quick cleanup, fix epJson input and CondenserEnteringTemp manager

amirroth committed:
Forgot to commit the unit test changes

amirroth committed:
Fix warnings

amirroth committed:
Merge remote-tracking branch 'origin/develop' into SPM

amirroth commented on performance_tests/15zonePVAV.idf:
```diff
@@ -7872,7 +7872,7 @@
 
   SetpointManager:MixedAir,
-    PACKAGED VAV WITH REHEAT_HeatC SAT Manager,  !- Name
+    PACKAGED VAV WITH REHEAT_HeatC SAT Manager MA,  !- Name
```
Duplicate SetPointManager names are no longer allowed.

mjwitte replied:
I changed the labels on this PR to include `IDDChange`, because this is changing input requirements even though the idd itself isn't changing. Will need transition for this.

amirroth commented on src/EnergyPlus/AirflowNetwork/src/Elements.cpp:
```diff
@@ -699,9 +699,6 @@ namespace AirflowNetwork {
 
-        // SUBROUTINE PARAMETER DEFINITIONS:
-        int constexpr CycFanCycComp(1); // fan cycles with compressor operation
-
```
This was not used.

amirroth commented on src/EnergyPlus/BoilerSteam.cc:
```diff
@@ -338,8 +338,7 @@ namespace BoilerSteam {
                 bool FatalError = false; // but not really fatal yet, but should be.
-                EMSManager::CheckIfNodeSetPointManagedByEMS(
-                    state, this->BoilerOutletNodeNum, EMSManager::SPControlType::TemperatureSetPoint, FatalError);
+                EMSManager::CheckIfNodeSetPointManagedByEMS(state, this->BoilerOutletNodeNum, HVAC::CtrlVarType::Temp, FatalError);
```
Unified a few redundant types to `HVAC::CtrlVarType`.

amirroth commented on src/EnergyPlus/DataEnvironment.hh:
```diff
@@ -62,6 +62,16 @@ struct EnergyPlusData;
 
 namespace DataEnvironment {
 
+    enum class GroundTempType
```
Made the different GroundTemp field groups into `std::array`s indexed by this `enum`.

amirroth commented on src/EnergyPlus/InputProcessing/InputProcessor.cc:
```diff
@@ -1377,28 +1381,28 @@ void InputProcessor::getObjectDefMaxArgs(EnergyPlusData &state,
-        json const alphas = legacy_idd["alphas"];
-        if (alphas.find("fields") != alphas.end()) {
-            NumAlpha += alphas["fields"].size();
+    if (auto found = legacy_idd.find("alphas"); found != legacy_idd.end()) {
```
Eliminate redundant lookups.

amirroth commented on src/EnergyPlus/InputProcessing/InputProcessor.hh:
```diff
@@ -127,7 +127,7 @@ public:
     bool getDefaultValue(EnergyPlusData &state, std::string const &objectWord, std::string const &fieldName, std::string &value);
 
-    std::string getAlphaFieldValue(json const &ep_object, json const &schema_obj_props, std::string const &fieldName);
+    std::string getAlphaFieldValue(json const &ep_object, json const &schema_obj_props, std::string const &fieldName, bool uc = true);
```
Not sure this param is really that useful. May delete later.

amirroth commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -197,37 +198,78 @@ constexpr std::array<std::string_view, static_cast<int>(SetPointManagerType::Num
+// No really, why?
+constexpr std::array<std::string_view, (int)AirTempType::Num> nodeTempTypeNamesUC = {
+    "NODEWETBULB", "NODEDRYBULB"};
+
```
Why are these called `OutdoorAirXBulb` and `NodeXBulb` when they both just mean `WetBulb`? Gratuitous differences result in gratuitous code.

amirroth commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -241,31 +283,43 @@ void ManageSetPoints(EnergyPlusData &state)
+    return (found != state.dataSetPointManager->spmMap.end()) ? found->second : 0;
+} // GetSetPointManagerIndex()
+        
 void GetSetPointManagerInputs(EnergyPlusData &state)
```
Not sure why this wrapper structure is needed.

amirroth commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -318,3730 +370,1075 @@ void GetSetPointManagerInputData(EnergyPlusData &state, bool &ErrorsFound)
-    NumZones = 0;
-    ZoneNum = 0;
-
-    cCurrentModuleObject = "SetpointManager:Scheduled";
```
As the kids say: Yeah, no.

amirroth commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -318,3730 +370,1075 @@ void GetSetPointManagerInputData(EnergyPlusData &state, bool &ErrorsFound)
+    int NumAlphas = 0; 
+    int NumParams = 0;
 
     state.dataInputProcessing->inputProcessor->getObjectDefMaxArgs(state, "NodeList", NumParams, NumAlphas, NumNums);
```
Not 100% sure how to read extensibles epJson style. Any examples of this?

mjwitte replied:
> Not 100% sure how to read extensibles epJson style. Any examples of this?

Here's one example.
https://github.com/NREL/EnergyPlus/blob/6cccca956da1ad4687ab5e184c2e1d5cfe79d9cf/src/EnergyPlus/DataZoneEquipment.cc#L1243-L1275

amirroth commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -318,3730 +370,1075 @@ void GetSetPointManagerInputData(EnergyPlusData &state, bool &ErrorsFound)
+#define PRESERVE_IDF_ORDER        
+#ifdef PRESERVE_IDF_ORDER
+        
+        // Try to read these in IDF order (for now, can revert to simpler begin(), ++, end() iterator later)
```
If you want an example of how to read epJson objects (useful for tracking down diffs), here it is. Same approach is used in `getObjectItem`.

amirroth commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -318,3730 +370,1075 @@ void GetSetPointManagerInputData(EnergyPlusData &state, bool &ErrorsFound)
+            SPMBase *spm = nullptr;
+
+            // Create a SetPointManagerObject of the right child type
+            switch (type) {
```
Create the right type of child-class. In Python, this would be a one liner.

amirroth commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -318,3730 +370,1075 @@ void GetSetPointManagerInputData(EnergyPlusData &state, bool &ErrorsFound)
+            case SPMType::MZHeatingAverage:
+            case SPMType::FollowOutsideAirTemp:
+            case SPMType::FollowGroundTemp: {
+                spm->minSetTemp = ip->getRealFieldValue(fields, props, "minimum_setpoint_temperature");
```
Shared code.

amirroth commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -318,3730 +370,1075 @@ void GetSetPointManagerInputData(EnergyPlusData &state, bool &ErrorsFound)
-        state.dataSetPointManager->AllSetPtMgr(AllSetPtMgrNum).NumCtrlNodes = state.dataSetPointManager->DualSchSetPtMgr(SetPtMgrNum).NumCtrlNodes;
-    }
+            // SetpointManager:Scheduled:DualSetpoint
+            case SPMType::ScheduledDual: {
```
Load `SetPointManager`-specific fields.

amirroth commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -318,3730 +370,1075 @@ void GetSetPointManagerInputData(EnergyPlusData &state, bool &ErrorsFound)
+            } // switch (spm->type)
+
+            // Load control node list
+            // Do this at the end to preserve node order
```
Just doing this once instead of 20 times saved 400 lines of code.

amirroth commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -5876,222 +2532,34 @@ void SimSetPointManagers(EnergyPlusData &state)
-    // The System Node Reset Humidity Setpoint Managers
-    for (SetPtMgrNum = 1; SetPtMgrNum <= state.dataSetPointManager->NumSystemNodeResetHumSetPtMgrs; ++SetPtMgrNum) {
-        state.dataSetPointManager->SystemNodeResetSetPtMgr(SetPtMgrNum + state.dataSetPointManager->NumSystemNodeResetTempSetPtMgrs).calculate(state);
+    for (auto *spm : state.dataSetPointManager->spms) {
```
Ahhhhhhhhhh.

amirroth commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -7976,81 +3636,68 @@ void DefineIdealCondEntSetPointManager::calculate(EnergyPlusData &state)
+    auto &dspm = state.dataSetPointManager;
+    
     // SUBROUTINE LOCAL VARIABLE DECLARATIONS:
-    auto &CondWaterSetPoint = state.dataSetPointManager->CondWaterSetPoint;
```
Bad use of shortcut refs.

amirroth commented on src/EnergyPlus/SetPointManager.hh:
```diff
@@ -66,61 +68,40 @@ struct EnergyPlusData;
 
 namespace SetPointManager {
 
-    enum class CtrlNodeType
```
Moved to `DataHVACGlobals`.

amirroth commented on src/EnergyPlus/SetPointManager.hh:
```diff
@@ -955,240 +505,77 @@ namespace SetPointManager {
 
     bool InitSetPointManagersOneTimeFlag = true;
     bool InitSetPointManagersOneTimeFlag2 = true;
-    Real64 DCESPMDsn_EntCondTemp = 0.0;
```
Most of these were actually local variables.

amirroth commented on tst/EnergyPlus/unit/HVACControllers.unit.cc:
```diff
@@ -133,16 +133,19 @@ TEST_F(EnergyPlusFixture, HVACControllers_ResetHumidityRatioCtrlVarType)
     GetSetPointManagerInputs(*state);
     // check specified control variable type is "HumidityRatio"
-    ASSERT_TRUE(compare_enums(SetPointManager::CtrlVarType::HumRat, state->dataSetPointManager->AllSetPtMgr(1).CtrlTypeMode));
+    ASSERT_EQ((int)HVAC::CtrlVarType::HumRat, (int)state->dataSetPointManager->spms(1)->ctrlVar);
```
This is better than `compare_enums`. Just use this from now on.

jmarrec replied:
I disagree. IMHO The custom function is clearer, and doesn't require casting, so you don't have to remember how to do it properly. But that's not even the real point here.

The point is that compare_enums will not let you shoot yourself in the foot. You **cannot** compare different enums with it, it will not compile.

If your beef is with the way the error message is structured, we can work on that to improve it.

---

And I'd avoid C-like cast, so at least it should be a static_cast<int>, which gets verbose.

An alternative would be define the C++23 std::to_underlying: https://en.cppreference.com/w/cpp/utility/to_underlying

That's dead simple

```
// std::to_underlying only in C++23
template <class T>
inline constexpr typename std::underlying_type_t<T> to_underlying(T val) noexcept {
  return static_cast<typename std::underlying_type<T>::type>(val);
}
```

But it'd still won't ensure you use the right types

Demo: https://godbolt.org/z/57PYfrs1Y

jmarrec replied:
How about we figure a way to have  `EXPECT_ENUM_EQ` and a `EXPECT_ENUM_NE` macros.

You could use them like

```c++
    auto v1 = EnumClass::val1;
    EXPECT_ENUM_EQ(v1, EnumClass::val2);
    EXPECT_ENUM_NE(v1, EnumClass::val1);
```

And they'd format to 

```shell
/path/tofile.cpp:107: Failure
In comparing enums of type 'EnumClass', Expected equality of these values:
  v1
    Which is: 0
  EnumClass::val2
    Which is: 1

/path/tofile.cpp:110: Failure
In comparing enums of type 'EnumClass'
  Expected: (v1) != (EnumClass::val1)
  Actual: both are 0
```

or maybe just

```
/path/tofile.cpp:107: Failure
Expected equality of these values:
  v1
    Which is: EnumClass(0)
  EnumClass::val2
    Which is: EnumClass(1)

/path/tofile.cpp:110: Failure
Expected: (v1) != (EnumClass::val1)  actual: both are EnumClass(0)
```

I'd be happy to work on that!

amirroth replied:
My beef is not with the error message, it's with the `compare_enums(X1, X2, false)` "gotcha", so I like the `EXPECT_ENUM_EQ`/`EXPECT_ENUM_NE` solution. We don't have to do it in this PR. I will revert these changes and we can do this separately. Thanks.

amirroth replied:
I reverted.

amirroth committed:
Revert compare_enum changes

amirroth committed:
Merge remote-tracking branch 'origin/develop' into SPM

rraustad commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -241,31 +283,43 @@ void ManageSetPoints(EnergyPlusData &state)
+{
+    auto found = state.dataSetPointManager->spmMap.find(Name);
+    return (found != state.dataSetPointManager->spmMap.end()) ? found->second : 0;
+} // GetSetPointManagerIndex()
```
So this is the reason all SPMs must have a unique name? Not that having unique names is a bad thing, or that users should want to have unique names across all SPMs, just that a name/type would always be unique if a unique name is used within each class. I am only thinking about this because of UnitarySystem now managing more parent object classes and whether there is anything I can do about not forcing the requirement of unique names (and the new getIndex function added to UnitarySystem in #10452). I think I am landing on unique names being a good thing going forward. Why would you name a PTAC the same as a CoilSystem? or a scheduled SPM the same as a mixed air SPM? In fact, it may be a good thing to force unique object names for every object in E+.

amirroth replied:
Well, there are at least three things going on here. First, previously name uniqueness was not forced on SPMs at all. Not within an SPM sub-class, not across sub-classes. And I am guessing this is because some created a situation in which SPM input data was read twice and couldn't think of a way to resolve this issue other than disabling the check.

Second, I don't think we actually need to enforce name uniqueness within the instances of a certain IDF object because epJSON does that already. It uses the object names as keys and you can't have duplicate keys.

Third, the "convention" within EnergyPlus right now is to enforce name uniqueness within a "parent" class for lack of a better term. So, there is name uniqueness across all curves, and all schedules, and all chillers, etc. This is what "Global Names" is used for and this is what I did here. We currently do not enforce name uniqueness across an entire IDF file. We can start to do that, I am not sure how much that will help. Name uniqueness within the same class hierarchy is helpful because ultimately you are going to want to handle members of that hierarchy in a uniform way (so you can delete 65% of the code!) and that means storing them in the same data structures. If you have multiple objects of the same name in the same data structure, things can get a bit messy/verbose because then you need a secondary key everywhere to tell one from the other. But you're not going to run into that problem for objects that are not in the same class hierarchy. Having a Schedule and a Fan with the same name is not a problem internally because those two objects will never be handled by the same piece of code.

jmarrec commented on src/EnergyPlus/SetPointManager.cc:
```diff
@@ -138,32 +137,34 @@ using namespace Curve;
-                                                                                                "MINIMUMMASSFLOWRATE"};
-
-constexpr std::array<std::string_view, static_cast<int>(ControlStrategy::Num)> strategyNamesUC = {
+constexpr std::array<std::string_view, (int)HVAC::CtrlVarType::Num> ctrlVarTypeNames = {
```
Im curious why change the static_cast to a C-style cast?

C++ style cast are checked by the compiler, more restrictive, and can be searched for easily.

(Never use a C-style cast in C++ is my rule)

amirroth replied:
Your opinion is more than welcome. I don't mark you as a reviewer because I figured you have enough other things to do (as do I, but that's a different story), but if you have the bandwidth and interest.

As for this, this is the only idiom (enum as an array index) where I use C-style casts, and it's mostly for readability. My presumption is that all enums are implemented as `int`s by the compiler. In my 30+ years of programming, I have never seen them implemented any other way, so the assumption is that this upcast is safe even without a check. Furthermore, if you are using an `enum` as a `std::array` index, you are already explicitly saying that you are using the `enum` as an it.

Note, I agree with you about the `EXPECT_EQ` thing. That's a different idiom, and using int casts there was a bridge too far. But I don't see a problem with them here in this idiom.

jmarrec replied:
I understand the rationale.

---

For completeness, The default is that C++ `enum class` uses integer, C enums are unsigned. In other words:

```
enum e1 {}; <=> enum e1: unsigned {}
enum class e2 {};  <=> enum e2: int (};
```

If you use the enum value for indexing into a std::array (and you can only do that because we do not have the related conversion compiler warnings enabled), you're actually saying `myarray[static_cast<std::size_t>(myEnumValue)]`. 

Which is probably fine in E+ since we typically go from Invalid = -1 to something small anyways, except if you were use the Invalid = -1 we have, and do `myarray[MyEnum::Invalid]`. static_cast<size_t>(-1) gives 18,446,744,073,709,551,616, which is definitely going to raise an out-of-bounds array indexing for our enums, but in a debug build only. In release, this is UB.

amirroth replied:
There are actually two kinds of `enum`s in EnergyPlus. Those that come from the IDF file, and those that are used only internally.  The ones used internally should never be set to `Invalid` and probably don't even need an `Invalid` value. The ones that come from EnergyPlus probably also don't need an `Invalid` value either because @rraustad loves to remind me, those get trapped in the epJson parser and never get to EnergyPlus proper.

You can go ahead and enable the compiler warnings. In fact, I see that somehow a `-w` snuck into `flags.cmake` in `energypluslib.dir` which you probably want to get rid of while you are at it.

jmarrec commented on src/EnergyPlus/SetPointManager.hh:
```diff
@@ -136,635 +117,269 @@ namespace SetPointManager {
-        }
+        int schedNum = 0;
 
         void calculate(EnergyPlusData &state);
```
```suggestion
        void calculate(EnergyPlusData &state) override;
```

Same comment applies to all of them deriving from SPMBase.

Your code works, but marking it override has two advantages:

* It's clearer that's what your doing, and clearer is good.
* The compiler will be nice enough to yell at you if you mess up the function signature and there's no such parent function to override.
    *  (I do realize in this specific case since your parent virtual function is a pure virtual one, so if you messed it up the compiler would still yell at you, but still)

amirroth replied:
TIL why it's good to use `override`. It always seemed nonsensical to me. Thanks!

jmarrec commented on src/EnergyPlus/SetPointManager.hh:
```diff
@@ -66,61 +68,40 @@ struct EnergyPlusData;
         MinTemp,
         Num
     };
+        
```
```suggestion

```

Trailing whitespaces in that file.

jmarrec commented on src/EnergyPlus/SetPointManager.hh:
```diff
@@ -955,240 +505,77 @@ namespace SetPointManager {
-    Real64 TotEnergy = 0;         // Total energy consumptions at this time step
-    Real64 TotEnergyPre = 0;      // Total energy consumptions at the previous time step
-
+    Array1D<SetPointManager::SPMBase *> spms;
```
I'm pretty sure you picked a raw pointer on purpose after consideration, but in the off chance not, I'd prefer using `std::unique_ptr<SPMBase>`. (And maybe a std::array or EPVector to contain it?)

jmarrec replied:
"But performance" you may say. well,

https://quick-bench.com/q/9xIjZoQqSyiz-uhlVFTFpN7fsn0

Locally I get more similar results, but still, my point is that this does not hinder performance.

![image](https://github.com/NREL/EnergyPlus/assets/5479063/3a61ba2a-0cc9-4304-ba31-a7d9f1b659f2)

amirroth replied:
> I'm pretty sure you picked a raw pointer on purpose after consideration, but in the off chance not, I'd prefer using `std::unique_ptr<SPMBase>`. (And maybe a std::array or EPVector to contain it?)

I did use a raw pointer on consideration. What is the consideration for `std::unique_pointer`?EnergyPlus is not a traditional scientific program, but it does have the same simple memory management  pattern as scientific programs do. You allocate everything at the beginning, then you compute, then you deallocate. There's no need for reference counting and garbage collection in that model. Just add `delete` loops to `clear_state`. To me, `std::unique_pointer` is just one more box in C++-library bingo.

Also, I'm out on `EPVector`. I no longer see the point and in fact I am convinced that a container that can be accessed as both a 0-based and 1-based array is only a source of bugs. I could change to `std::vector` and I have done that in some modules that are more self-contained than `SPM` (e.g., `OutputProcessor`). It will probably take another day or so to dig up the corner cases that arise. Is that worth it for this PR?

amirroth replied:
> "But performance" you may say. well,
> 
> https://quick-bench.com/q/9xIjZoQqSyiz-uhlVFTFpN7fsn0
> 
> Locally I get more similar results, but still, my point is that this does not hinder performance.
> 
> ![image](https://private-user-images.githubusercontent.com/5479063/336021802-3a61ba2a-0cc9-4304-ba31-a7d9f1b659f2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc0MTk1MjQsIm5iZiI6MTcxNzQxOTIyNCwicGF0aCI6Ii81NDc5MDYzLzMzNjAyMTgwMi0zYTYxYmEyYS0wY2M5LTQzMDQtYmEzMS1hN2Q5ZjFiNjU5ZjIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MDNUMTI1MzQ0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzEwMzhkMTdlMGRlYTI5NjkwNDM3YmM1YTI2NzNkY2ZlYWFiY2Y2MWVkYTAzNmJhN2VjNWVlMDFhZWRhZjlhMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Zd8UHWPfAHJRCXWpyd5CyZiqdzisd9rSoboFIuxQkaQ)

I will not say "but performance". I will say "C++ bingo".

jasondegraw replied:
Regarding `EPVector`, the original intent was as a shim to move toward `std::vector` and `std::array` and not as something we should keep around forever. With `EPVector`, we wanted something that would work well enough with `Array1D` (which allows 0-based and 1-based indexing and is the reason that `EPVector` does it) that switching wouldn't break everything.

New code should definitely use `std::vector/array` unless there's a compelling reason to go another direction. At the time that `EPVector` was introduced, there were a lot of problematic cases (slicing etc.) that are less present now. Anything that's an `EPVector` now should (hopefully) just need the indexing addressed in order to be switched.

amirroth replied:
I agree. I think this is why I changed my mind about it. I was pro `EPVector` earlier because there was just a lot of `Array1D` madness that we needed to move away from. But now that we've moved away from a lot of it, the 0-based/1-based duality has become somewhat of a bug rather than a feature for external references (I was bitten by it a few times during the `Fans` transition). FWIW I actually think that Array1D also allows mixed indexing to some degree and we should disable it. 

I also agree that we need to move to 0-based indices incrementally and I was thinking of doing that here, but it took me a little while to actually work the kinks out of this module and by the time I was done I was too tired to deal with that type of change. I will probably go back and do it in a separate pass.

jasondegraw replied:
There used to be way too much slicing going on, that's for sure. Let's not add any new `Array1D` objects if we can help it. With Fortran indexing being what it is (negative indices and all that), disabling either `operator()` or `operator[]` on `Array1D` might be harder than just moving to `EPVector` and then on to `std::vector`.

amirroth replied:
I was thinking of disabling `operator []` on it. I don't think it's used in more than a handful of places, maybe I'm wrong.

jmarrec commented on src/EnergyPlus/SetPointManager.hh:
```diff
@@ -955,240 +505,77 @@ namespace SetPointManager {
-    Real64 TotEnergyPre = 0;      // Total energy consumptions at the previous time step
-
+    Array1D<SetPointManager::SPMBase *> spms;
+    std::map<std::string, int> spmMap;
```
Seems like this is only used in GetSetPointManagerIndex, which you could rewrite to just scan the `spms` array. 

I'm doubtful the std::map is improving the performance.

amirroth replied:
map is probably not improving performance for SPM's (there aren't that many SPMs), but it should for things like nodes, surfaces, and other objects for which there are a lot of instances. You have to commit to the bit.

Also, the map is pulling double duty here. It's both for lookups and for finding duplicate names. In other modules, we use a map for finding duplicate names, but then do linear search to find indices ... which is two kinds of dumb in my opinion.

jmarrec commented on src/EnergyPlus/SetPointManager.hh:
```diff
@@ -955,240 +505,77 @@ namespace SetPointManager {
-        NoDeepGroundTempObjWarning = true;
-        NoFCGroundTempObjWarning = true;
+
+        for (int iSPM = 1; iSPM <= (int)spms.size(); ++iSPM) {
+            delete spms[iSPM];
+        }
+        spms.deallocate();
```
Loop Would not be necessary if using a std::unique_ptr

amirroth replied:
`std::unique_ptr` would be unnecesary if using loop. Why use the more complicated feature when the simpler one will do?

jmarrec replied:
I've read your response above and I get it (https://github.com/NREL/EnergyPlus/pull/10537#discussion_r1624433855)

---

It's getting philosophical perhaps, but my opinion is that unique_ptr (and other smart pointers) have been created for a reason. They are there to avoid the numerous leaks that were common currency in older codebases. Your argumentation is one that's shared by a lot of people though: why would I need this new feature when my code is already 100% correct and I'm managing memory perfectly? Which typically goes with "why would I pay for overhead when I know what I'm doing" (and overhead there can be indeed!). I can understand both, perhaps I'm just not confident enough to assume I'll never do something stupid and leak. I also don't feel that this is more complicated in this case, and the fact that I no longer have to worry about memory removes a cognitive burden so it's a net win for me _personally_.

amirroth replied:
I don't want to sound like the C-blowhard that I have been accused of being. I am fine with using C++ (and standard library) features when they are needed. There are many C++ features I wish we would use more, starting with ... you know ... polymorphism! I am less fine with using C++ features (or any features) just because it feels good to do it. EnergyPlus is not a venue for people to experiment with cool looking C++ features (if you ever suspect me of doing such a thing, please call me on it!).

You find the additional layer of `std::unique_ptr` to be comforting. I find it to be annoying especially when implementing polymorphism and needing to use `dynamic_cast`. To me, having a `delete` loop in a destructor is simpler and more comforting than toting the `unique_ptr` wrapper around, no matter how thin it is. 

Maybe I spent too many years programming in C and just got acclimated to doing explicit memory management so I don't find it to be a cognitive burden.  I will say that I believe that the EnergyPlus team in general has too little familiarity with memory management and memory layout and that this manifests as numerous tiny inefficiencies throughout the code (misuse of references creating either unnecessary copies or unnecessary indirection, poor data structure design) whose cumulative effect is probably no more than 10-15% and whose chief impact is to upset me every time I come across one. TL;DR I find having to think about memory management (to the point where you get good at it and it becomes a mindless habit) to be a feature. I would say that I am surprised that given how much development you have done you don't feel the same way. I suspect that some of this has to do with the fact that you've done a lot of development in code that does a lot of dynamic memory management (e.g., OpenStudio) where the cognitive load of memory management is far greater and features like `unique_ptr` and `shared_ptr` are much more useful. Incidentally, this is one reason why I would balk at combining EnergyPlus and OpenStudio into a single code base. I just think that EnergyPlus' current memory management model of allocate/compute/tear-down is too valuable of a simplification to abandon for uncertain gains.

jmarrec reviewed (commented):
Couple of nitpicks (I know, no one asked for my opinion, sorry).

Myoldmopar commented:
Lots of fantastic stuff here.  And my god, NET -6000 lines of code is amazing.  

I appreciate your commentary @jmarrec, some of which led to commits and other things which are tabled for later.

I am happy to apply Clang Format and get that cleaned, but as @mjwitte noted, this will need a transition rule.  It will need to, at a minimum, alert the user about the unique name requirement, and at a maximum, actually rename things.  I'm guessing the former is sufficient.  @mjwitte do you know of an example of that already in place in one of the transition versions?  Or should I just try it out from scratch?

amirroth commented:
> Lots of fantastic stuff here. And my god, NET -6000 lines of code is amazing.
> 
> I appreciate your commentary @jmarrec, some of which led to commits and other things which are tabled for later.
> 
> I am happy to apply Clang Format and get that cleaned, but as @mjwitte noted, this will need a transition rule. It will need to, at a minimum, alert the user about the unique name requirement, and at a maximum, actually rename things. I'm guessing the former is sufficient. @mjwitte do you know of an example of that already in place in one of the transition versions? Or should I just try it out from scratch?

I have one more quick commit based on @jmarrec comments.

There is already a duplicate name error printed.

amirroth committed:
SetPointManager.hh

Myoldmopar commented:
> There is already a duplicate name error printed.

You are referring to the EnergyPlus code itself failing, right?  Should we also alert the user during the transition process?  This is really a change of behavior from 24.1 to 24.2, so I think the idea is to let them know about the change while they are updating their input files.  Of course if someone makes a brand new file in 24.2, we'd also warn them that it is invalid when the execute that file.  

I'm not strongly opinionated here, but happy to chip in on a transition warning if it is warranted.

amirroth commented:
> > There is already a duplicate name error printed.
> 
> You are referring to the EnergyPlus code itself failing, right? Should we also alert the user during the transition process? This is really a change of behavior from 24.1 to 24.2, so I think the idea is to let them know about the change while they are updating their input files. Of course if someone makes a brand new file in 24.2, we'd also warn them that it is invalid when the execute that file.
> 
> I'm not strongly opinionated here, but happy to chip in on a transition warning if it is warranted.

Yes, I was referring to EnergyPlus itself. I don't know about the other thing. If anything, we should probably have been alerting the user to the fact that duplicate names were **not** being check-for previously since that seems to be the "rule" for pretty much every other "family" of objects in EnergyPlus.

amirroth committed:
Merge remote-tracking branch 'origin/develop' into SPM

amirroth committed:
Merge remote-tracking branch 'origin/develop' into SPM

mjwitte commented:
> I am happy to apply Clang Format and get that cleaned, but as @mjwitte noted, this will need a transition rule. It will need to, at a minimum, alert the user about the unique name requirement, and at a maximum, actually rename things. I'm guessing the former is sufficient. @mjwitte do you know of an example of that already in place in one of the transition versions? Or should I just try it out from scratch?

I can't think of a similar transition. The good thing is that SPM names are never used elsewhere in the input except possibly for the two output variables for SetpointManager:WarmestTemperatureFlow.

jmarrec committed:
SPM - Fortran Transition - throw if SPM names are not unique

jmarrec commented:
@Myoldmopar I made a working implementation of detecting non unique SPM names accross all types and throwing a fatal if not in #10542
(Don't ask me why I did that, I don't know)

amirroth commented:
> @Myoldmopar I made a working implementation of detecting non unique SPM names accross all types and throwing a fatal if not in #10542
> (Don't ask me why I did that, I don't know)

Because you’re a sicko, that’s why.

Myoldmopar committed:
Merge pull request #10542 from NREL/SPM-transition

SPM - Fortran Transition - throw if SPM names are not unique

Myoldmopar committed:
Apply Clang Format

Myoldmopar commented:
Alright, thanks to @jmarrec this now has the transition rules in place to enforce (warn) uniqueness during transition workflows.  I just applied Clang Format, so CI should be clean.  There has been some good conversation here about the overall architecture and philosophy, and that can continue for a bit while we wait on things to get confirmed.

amirroth commented:
> Alright, thanks to @jmarrec this now has the transition rules in place to enforce (warn) uniqueness during transition workflows. I just applied Clang Format, so CI should be clean. There has been some good conversation here about the overall architecture and philosophy, and that can continue for a bit while we wait on things to get confirmed.

We're not discussing architecture per se, we're discussing use of various C++ features and idioms. Specifically:

- Use of `EXPECT_TRUE(compare_enums(e1, e2))` vs.  `EXPECT_EQ((int)e1, (int)e2)`. I reverted my changes on this one and agree that we should preserve `enum`-typed comparison here although maybe with a cleaner template that doesn't have the `compare_enums(e1, e2, false)` gotcha.
- Benefit of `override` keyword. Added.
- Use of unchecked `(int)` "upcasts" vs. checked `static_cast<int>` "upcasts" in use of `enums` as `std::array` indices. Not backing down on this one at this point.
- Use of raw pointers and `delete` loops in `clear_state` vs. `unique_ptr` wrappers. Not backing down on this one either right now. I've made my thoughts clear about this feature. I understand its utility in general but [EnergyPlus specifically doesn't need it](https://github.com/NREL/EnergyPlus/wiki/%22Smart-Pointers%22%3A-Why-Are-You-Using-Them%3F). At best, it's an emotional support idiom. At worst, it's C++-library bingo.
- Continued use of `Array1D` vs. transition to `std::vector` when these refactoring changes are made. This is a gray one. I've been transitioning to `vector` when  I feel like the module is sufficiently self contained and there aren't too many external indices (e.g., `OutputProcessor`, although @JasonGlazer found one last week!). If I think that I may have to chase down a bunch of `if (xNum > 0)` tests then I leave it as `Array1D` (e.g., `Fans`). The situation with SPM fell in the middle and towards the end I just got too tired to tackle this part.

jmarrec committed:
Merge branch 'develop' into SPM and resolve conflicts

jmarrec commented:
I merged develop in and resolved the conflicts due to the merge of

* #10550

Myoldmopar commented:
OK, I think this one is ready to go.  I'll let CI wrap up since it is almost done, but it looks clean, just BND diffs.  Speak up quick if there is anything else here!

Myoldmopar commented:
OK fine, merging!  Thanks @amirroth.  And reviewers.
